text/3714-macro-fragment-fields.md

```diff
@@ -0,0 +1,182 @@
+
+The macros-by-example system is powerful, but sometimes difficult to work with.
+In particular, parsing complex parts of Rust syntax often requires carefully
+recreating large chunks of the Rust grammar, in order to parse out the desired
+pieces. Missing or incorrectly handling any portion of the syntax can result in
+not accepting the same syntax Rust does; this includes future extensions to
+Rust syntax that the macro was not yet aware of. Higher-level fragment
+specifiers are more robust for these cases, but don't allow extracting
+individual pieces of the matched syntax.
+
+This RFC introduces a mechanism to use high-level fragment specifiers while
+still extracting individual pieces of the matched syntax.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing macros by example, and using certain high-level fragment
+specifiers, you can use the syntax `${matched_name.field_name}` to extract
+specific "fields" of the matched syntax. This allows you to use the Rust parser
+for those high-level fragments, rather than having to recreate parts of the
+Rust grammar in order to extract the specific pieces you want. Fields evaluate
+to pieces of Rust syntax, suitable for substitution into the program or passing
+to other macros for further processing.
+
+For example, the fragment `:adt` parses any abstract data type supported by
+Rust: struct, union, or enum. Given a match `$t:adt`, you can obtain the name
+of the matched type with `${t.name}`:
+
+```rust
+macro_rules! get_name {
+    ($t:adt) => { println!("{}", stringify!(${t.name})); }
+}
+
+fn main() {
+    let n1 = get_name!(struct S { field: u32 });
+    let n2 = get_name!(enum E { V1, V2 = 42, V3(u8) });
+    let n3 = get_name!(union U { u: u32, f: f32 });
+    println!("{n3}{n1}{n2}"); // prints "USE"
+}
+```
+
+An attempt to access a field that doesn't exist will produce a compilation
+error on the macro definition, whether or not the specific macro rule gets
+invoked.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Fragment fields may be used in a macro transcriber anywhere the corresponding
+fragment name could be used.
+
+Fragment fields typically follow the same rules for repetition handling as the
+corresponding fragment (e.g. being used at the same level/kind of repetition).
+However, fragment fields that contain multiple items require one additional
+level of repetition; see the `param` field of `:fn`, below.
+
+This RFC introduces the following new fragment specifiers, with specified fields:
+
+- `:fn`: A function definition (including body).
+  - `name`: The name of the function, as an `ident`.
+  - `param`: The parameters of the function, presented as though captured by a
+    level of `*` repetition. For instance, you can write `$(${f.param}),*` to
+    get a comma-separated list of parameters, or `$(other_macro!(${f.param}))*`
+    to pass each parameter to another macro.
```

Comment:
Shouldn't we have two lists, one with the parameter patterns (typically names), and one with the types?

Or at least this should clarify that a `param` has the shape `pat: type`, and explain why this is provided as a single unit rather than being split up further.

Reply:
:thinking: there are some exotic parameters like `self`/`&self`/`&mut self` and `...` which shouldn't or can't be split into two lists.

and if the function declares a parameter as `#[some_inert_attr] mut x: T` should the attribute and `mut` be included into the `param`?

Reply:
`mut` is part of the "pattern" (maybe there's a better name for that)... but the rest are good questions I don't have a clear answer to.

Reply:
I think `&self` could be potentially split into `self` and `&Self`?

Reply:
also then, what about `...`?

Reply:
> Shouldn't we have two lists, one with the parameter patterns (typically names), and one with the types?
> 
> Or at least this should clarify that a `param` has the shape `pat: type`, and explain why this is provided as a single unit rather than being split up further.

There's an item in future work to add a matcher for a single function parameter, which would have fields for the pattern and type. One reason I deferred this is that people may expect *name* and type, and the vast majority of cases will have a name, but not every case will. So, better handling of parameters needs some discussion of how to handle fields that won't always exist, perhaps with some kind of conditional handling. I want to defer all of that.

It's forwards-compatible to add fields to a matcher.